### ISSUE TITLE
PageView stops responding to input on a particular page number, on some devices

### ISSUE THREAD

**Issue by ttencate** [flutter/flutter#65747]
I've hesitated for a long time to report this, because I can't reproduce it myself and therefore can't create a minimal reproduction project. Heck, I can't even be sure that the bug is in Flutter and not in my own code, although I can't imagine how that would be. Fortunately my [app](https://papageno.app) is [open source](https://github.com/ttencate/papageno), so I'm hoping one of the Flutter devs has more luck.

## Steps to Reproduce

1. Install https://play.google.com/store/apps/details?id=com.frozenfractal.papageno.  
   Or clone https://github.com/ttencate/papageno and run `flutter run` in the `app` directory.
2. Tap "Use current location" and "Create course" to create a new course for your location.
3. Tap "Start practice".
4. Keep tapping bird names and tapping to proceed to the next question.
5. At some point, the app _might_ stop responding to taps. Play/pause, tapping bird names, and tapping the menu button all don't work. Swiping back to the previous question still works, but if you tap/swipe forwards again, you get back to the non-working state.
6. If you hit the Back button to abort the practice session and start a new one, the app stops responding to taps on the _same_ question number as before. (Question numbers aren't shown, you'll have to keep count or look at the progress bar at the bottom.)

Do you get through all the 24 birds and see your percentage score? Bummer: the issue is not reproducible on this device. This is the case for the majority of devices.

My suspicion is that there is some floating point issue at play here, which causes the `PageView` to get stuck at page 18.9999 instead of proceeding to page 19. That would explain why it's not responding to taps (a tap while halfway between pages merely stops the current scroll animation) but why swiping still works fine.

## Device info

This happens only on some devices, but consistently. Note: I'm _assuming_ it has to do with the device, but I can't rule out that it's caused by some other factor.

* **Motorala Moto G6 Plus** with unknown Android version: always gets stuck on the 20th bird.
* **Samsung Galaxy S10** with Android 10: always gets stuck on the 6th bird.

The issue is not reproducible on my own devices (FairPhone 2 with Android 7.1). I tried to match the specs of the Moto G6 in the Android emulator, but wasn't able to reproduce it there either.

## Code info

Roughly, the widget tree looks as follows:

```
Column
  Expanded
    PageView.builder(physics: PageScrollPhysics)
      itemBuilder -> QuestionScreen
        GestureDetector
          Scaffold
            appBar: AppBar
            body: Column
              ...
    LinearProgressIndicator
```

So each question is a page in a `PageView`. To enable the "Tap to continue" behaviour, the entire page is wrapped in a `GestureDetector` whose `onTap` is conditionally set to a non-`null` handler. The handler just [calls](https://github.com/ttencate/papageno/blob/cf4723a95ccc28e01f1f7fd09c35c73ff18927e8/app/lib/screens/quiz_page.dart#L170) `_pageController.animateToPage` to go to the next page.

The `PageView` has its `itemCount` set to `null` for the reasons mentioned in [this comment](https://github.com/ttencate/papageno/blob/cf4723a95ccc28e01f1f7fd09c35c73ff18927e8/app/lib/screens/quiz_page.dart#L108). Instead, the `PageView`'s `itemBuilder` explicitly returns `null` for pages that shouldn't yet be accessible.

From a previous version of the app, where the app bar was placed outside the `PageView`, I know that the app bar (specifically the menu button) _did_ keep responding. So it goes wrong somewhere within the `PageView` widget.

Because I don't have access to logcat output, I can't completely rule out that Flutter reported any errors, although I would hope that any unhandled exception would terminate the app, and not silently be swallowed. I've hooked up [my own exception logging](https://github.com/ttencate/papageno/blob/cf4723a95ccc28e01f1f7fd09c35c73ff18927e8/app/lib/main.dart#L39) for exceptions thrown from `async` functions, and we're not getting any of those.

## Attempted workarounds

I've tried increasing the `distance` and `velocity` tolerance on the `ScrollPhysics`, as you can see in [this commit](https://github.com/ttencate/papageno/commit/fff7516a4f240eb122a4d8135b3af2560f707d05), but it didn't help.

## Possibly related issues

* #63689 "Floating point issue in PageView extent calculation causes the next page to sometimes build"
* #60295 "PageView Builder is creating upcoming page asymetrically"
* #62209 "PageView controller's jumpToPage doesn't work properly on Android when jumping to a page after a new child was added, since 1.19.0-4.1.pre", which is on a newer version of Flutter, but @Hixie mentions "I don't really understand why this ever worked" so maybe it's relevant?
* #36876 "PageView doesn't always fully complete its nextPage() animation" in combination with #39984 "Pages in PageView are not clickable when scrolling" (which is as intended)
* On https://github.com/flutter/flutter/issues/60295#issuecomment-651040442 somebody also mentions behaviour that happens on specific page numbers.

<details>
  <summary>Logs</summary>

Because I can't reproduce the issue myself, I don't have access to logcat output. The log below is from my own logging, based on the [`logging`](https://pub.dev/packages/logging) package. The user answered question 4, proceeded to question 5 (probably by swiping, because we don't see output from the `onTap` handler), and then swiped back to question 4 to regain control and send me the log (using the log mailer built into the app).

```
FINER 2020-09-11 12:58:30.015707 Current page changed to 4
FINER 2020-09-11 12:58:30.654442 _QuestionScreenState.dispose() for question 3
FINER 2020-09-11 12:58:35.602292 _QuestionScreenState._choose(33194:Prunella modularis:Dunnock) for question 4
FINER 2020-09-11 12:58:35.602473 Answering question 4 with 33194:Prunella modularis:Dunnock
FINE 2020-09-11 12:58:35.602917 Updating species knowledge after Question(correctAnswer: 32166:Passer domesticus:House Sparrow, givenAnswer: 33194:Prunella modularis:Dunnock, isCorrect: false)
FINER 2020-09-11 12:58:36.843386 _QuestionScreenState.initState() for question 5 (Question(correctAnswer: 5163:Columba palumbus:Common Wood Pigeon, givenAnswer: null, isCorrect: null))
FINER 2020-09-11 12:58:36.845890 _QuestionScreenState.didChangeDependencies() for question 5
FINER 2020-09-11 12:58:36.892462 Current page changed to 5
FINER 2020-09-11 12:58:37.513565 _QuestionScreenState.dispose() for question 4
FINER 2020-09-11 12:58:39.219236 _QuestionScreenState.initState() for question 4 (Question(correctAnswer: 32166:Passer domesticus:House Sparrow, givenAnswer: 33194:Prunella modularis:Dunnock, isCorrect: false))
FINER 2020-09-11 12:58:39.219910 _QuestionScreenState.didChangeDependencies() for question 4
FINER 2020-09-11 12:58:39.242513 _QuestionScreenState.dispose() for question 4
FINER 2020-09-11 12:58:40.123237 _QuestionScreenState.initState() for question 4 (Question(correctAnswer: 32166:Passer domesticus:House Sparrow, givenAnswer: 33194:Prunella modularis:Dunnock, isCorrect: false))
FINER 2020-09-11 12:58:40.123654 _QuestionScreenState.didChangeDependencies() for question 4
FINER 2020-09-11 12:58:40.372878 _QuestionScreenState.dispose() for question 4
FINER 2020-09-11 12:58:41.794897 _QuestionScreenState.initState() for question 4 (Question(correctAnswer: 32166:Passer domesticus:House Sparrow, givenAnswer: 33194:Prunella modularis:Dunnock, isCorrect: false))
FINER 2020-09-11 12:58:41.795277 _QuestionScreenState.didChangeDependencies() for question 4
FINER 2020-09-11 12:58:41.893820 Current page changed to 4
FINER 2020-09-11 12:58:42.282141 _QuestionScreenState.dispose() for question 5
```

```
$ flutter analyze
Analyzing app...                                                        
No issues found! (ran in 7.9s)
```

```
$ flutter doctor -v
  ╔════════════════════════════════════════════════════════════════════════════╗
  ║ A new version of Flutter is available!                                     ║
  ║                                                                            ║
  ║ To update to the latest version, run "flutter upgrade".                    ║
  ╚════════════════════════════════════════════════════════════════════════════╝


[✓] Flutter (Channel stable, v1.17.1, on Linux, locale en_US.UTF-8)
    • Flutter version 1.17.1 at /home/thomas/flutter
    • Framework revision f7a6a7906b (4 months ago), 2020-05-12 18:39:00 -0700
    • Engine revision 6bc433c6b6
    • Dart version 2.8.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /opt/android-sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /opt/android-sdk
    • ANDROID_SDK_ROOT = /opt/android-sdk
    • Java binary at: /usr/lib/jvm/default/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_265-b01)
    • All Android licenses accepted.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2019.3)
    • IntelliJ at /usr/share/jetbrains-idea-ce
    • Flutter plugin version 44.0.3
    • Dart plugin version 193.6911.31

[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```

</details>

**Comment by darshankawar** [comment]
Hi @ttencate,
Thanks for a detailed report. I cloned your repo and tried to run, but getting below error on `fluter run`:

```
Invalid plugin specification moor_ffi.
Cannot find the `flutter.plugin.platforms` key in the `pubspec.yaml` file. An
instruction to format the `pubspec.yaml` can be found here:
https://flutter.dev/docs/development/packages-and-plugins/developing-packages#pl
ugin-platforms
```
Also, although you mentioned it's not replicable everytime on every device, I see that you've used a lot of 3rd party plugins in the app. Would it be possible for you to check by removing the 3rd party plugin code on the page numbers that stops responding on input ? Idea is to segregate the issue as much as possible to know it's not causing due to the plugins used.
Thanks.

**Comment by ttencate** [comment]
Thank you for checking this out so quickly @darshankawar!

Sorry about the issue with `moor_ffi`. The dependency on `moor` was on an experimental branch, but somehow ended up in the `pubspec.lock` file of the master branch. I just fixed it in `master`, so if you do `git pull`, it should be better.

I don't think that this issue is related to any plugin, because none of the plugins except `flutter_map` do anything with UI widgets, and `flutter_map` is not used on the screens that have the problem.

**Comment by darshankawar** [comment]
Hi @ttencate,
I was able to compile and install the app, but it gets stuck on launch screen and I see below and doesn't go past this screen:

![65747](https://user-images.githubusercontent.com/67046386/93169679-3dcdfa00-f743-11ea-9136-b4ae9a020b72.png)

Also see below log:

<details>
<summary>log</summary>

```
Restarted application in 3,029ms.
I/flutter (27373): INFO 2020-09-15 11:00:47.292996 Starting application  
E/SQLiteLog(27373): (11) Invalid MagicHeader!
E/SQLiteLog(27373): (11) lockBtree() error, rc: 26, printing out first page (size: 4096) of  DB /data/user/0/com.frozenfractal.papageno/databases/app.db
E/SQLiteLog(27373): (11) Page (1) content:
E/SQLiteLog(27373): (11)  - 76 65 72 73 69 6F 6E 20 68 74 74 70 73 3A 2F 2F 67 69 74 2D 6C 66 73 2E 67 69 74 68 75 62 2E 63
E/SQLiteLog(27373): (11)  - 6F 6D 2F 73 70 65 63 2F 76 31 0A 6F 69 64 20 73 68 61 32 35 36 3A 65 64 32 34 62 62 38 30 30 30
E/SQLiteLog(27373): (11)  - 37 64 66 35 30 36 39 32 39 65 34 32 31 66 34 33 66 64 37 35 32 62 39 61 61 63 30 65 36 32 65 32
E/SQLiteLog(27373): (11)  - 38 35 62 61 62 37 31 32 30 65 32 31 64 65 36 30 38 39 64 31 66 35 0A 73 69 7A 65 20 35 35 31 33
E/SQLiteLog(27373): (11)  - 32 31 36 0A 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
E/SQLiteLog(27373): (11)  - 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
I/chatty  (27373): uid=10187(com.frozenfractal.papageno) Sqflite identical 121 lines
E/SQLiteLog(27373): (11)  - 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
E/SQLiteLog(27373): (26) file is encrypted or is not a database
I/flutter (27373): SEVERE 2020-09-15 11:00:52.203086 uncaught future error DatabaseException(file is encrypted or is not a database (code 26): , while compiling: select * from species) sql 'select * from species' args []} #0      wrapDatabaseException (package:sqflite/src/exception_impl.dart:11:7)
I/flutter (27373): <asynchronous suspension>
I/flutter (27373): #1      SqfliteDatabaseFactoryImpl.wrapDatabaseException (package:sqflite/src/factory_impl.dart:78:7)
I/flutter (27373): #2      SqfliteDatabaseMixin.safeInvokeMethod (package:sqflite_common/src/database_mixin.dart:208:15)
I/flutter (27373): #3      SqfliteDatabaseMixin.txnRawQuery.<anonymous closure> (package:sqflite_common/src/database_mixin.dart:394:36)
I/flutter (27373): #4      SqfliteDatabaseMixin.txnSynchronized.<anonymous closure> (package:sqflite_common/src/database_mixin.dart:327:22)
I/flutter (27373): #5      BasicLock.synchronized (package:synchronized/src/basic_lock.dart:32:26)
I/flutter (27373): #6      SqfliteDatabaseMixin.txnSynchronized (package:sqflite_common/src/database_mixin.dart:323:33)
I/flutter (27373): #7      SqfliteDatabaseMixin.txnRawQuery (package:sqflite_common/src/database_mixin.dart:393:1

```

</details>

**Comment by ttencate** [comment]
Thank you for posting the log, which pinpoints the issue: the Git repository uses Git LFS for large files, so you need to install that first. Otherwise the database that is packaged with the app will be checked out as a Git LFS pointer file, not a real SQLite database file, causing this crash (proper error handling is still TODO).

You can install Git LFS using the instructions at https://git-lfs.github.com/. After running `git lfs install` successfully, you can run `git lfs pull` to fetch the actual files and update your existing working copy (or just clone it from scratch). If it worked correctly, the file `app/assets/app.db` should be 5513216 bytes in size.

Thank you so much for bearing with me. I really appreciate it.

**Comment by darshankawar** [comment]
@ttencate thanks. It would be really helpful if you could provide a reproducible code sample that we can simply run in IDE. Probably clone the feature and code on which the input stops working that involves PageView, similar in the line of what you mentioned earlier:

```
Column
  Expanded
    PageView.builder(physics: PageScrollPhysics)
      itemBuilder -> QuestionScreen
        GestureDetector
          Scaffold
            appBar: AppBar
            body: Column
              ...
    LinearProgressIndicator
```

**Comment by ttencate** [comment]
I will gladly do that, with the caveat that I cannot verify that it will actually reproduce the bug! I will try my best not to strip out any code that might be relevant.

The only thing I'm sure reproduces it is the version published on the Play Store, but as said, even that one works fine on the majority of devices. So maybe you can use the Play Store version to check whether it's reproducible on your particular test device(s) at all?

**Comment by darshankawar** [comment]
@ttencate I installed app from playstore on Samsung Galaxy A2Core device (8.1.0) and was able to complete the practice without any issues. I stopped / resumed along the way without any issues.

![65747](https://user-images.githubusercontent.com/67046386/93295973-e6479100-f80b-11ea-99ad-08095bede964.png)

But when I tried the same on Samsung S10+, I got stuck at 6th bird (as you mentioned) and couldn't tap anywhere. I pressed back button, quit the practice and started over but got stuck at same screen again.

![65747_s10](https://user-images.githubusercontent.com/67046386/93296522-062b8480-f80d-11ea-8cb9-67e4f91e7677.png)

So atleast on Samsung S10+ device, the issue is consistently replicable. Only thing now we would want is a minimal code sample that takes user towards the 6th bird screen (Pageview).

**Comment by ttencate** [comment]
Great to hear that you were able to repro it @darshankawar, and that it does seem to be consistently dependent on the device.

I've made a stripped down version that will hopefully exhibit the same behaviour. I tried to match the original controller code fairly closely, including when widgets get rebuilt, so it's still about 250 lines. I didn't throw away more for fear of removing something that's relevant to the issue. Simply do this to run it:

1. `flutter create flutterissue65747`
2. `cd flutterissue65747`
3. Replace `lib/main.dart` with the attached code.
4. `flutter run`

<details>
<summary>lib/main.dart</summary>

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

Future<void> main() async {
  runApp(App());
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      onGenerateTitle: (BuildContext context) => 'Flutter issue 65747',
      theme: ThemeData(),
      home: QuizPage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class Quiz {
  final int questionCount = 24;
  int availableQuestionsCount = 0;
  int answeredQuestionsCount = 0;

  bool get isComplete => answeredQuestionsCount >= questionCount;
}

class QuizController {

  Future<void> _questionsFuture;
  final Quiz _quiz = Quiz();
  final _quizUpdatesController = StreamController<Quiz>();

  QuizController() {
    _ensureQuestions();
  }

  Stream<Quiz> get quizUpdates => _quizUpdatesController.stream;

  void answerQuestion(int index) {
    if (index >= _quiz.availableQuestionsCount || index != _quiz.answeredQuestionsCount) {
      return;
    }
    _quiz.answeredQuestionsCount++;
    _notifyListeners();
    _ensureQuestions();
  }

  void dispose() {
    _quizUpdatesController.close();
  }

  void _ensureQuestions() {
    if (_questionsFuture != null ||
        _quiz.answeredQuestionsCount < _quiz.availableQuestionsCount ||
        _quiz.availableQuestionsCount >= _quiz.questionCount) {
      return;
    }
    _questionsFuture = () async {
      await Future.delayed(Duration(milliseconds: 2000));
      _quiz.availableQuestionsCount += 8;
      _notifyListeners();
      _questionsFuture = null;
    }();
  }

  void _notifyListeners() {
    _quizUpdatesController.add(_quiz);
  }
}

class QuizPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _QuizPageState();
}

class _QuizPageState extends State<QuizPage> {
  final _quizController = QuizController();
  final _pageController = PageController(initialPage: 0);
  int _currentPage = 0;

  @override
  void dispose() {
    _pageController.dispose();
    _quizController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return StreamBuilder<Quiz>(
      stream: _quizController.quizUpdates,
      builder: (context, snapshot) {
        final quiz = snapshot.data;
        final theme = Theme.of(context);
        final drawer = Drawer(
          child: ListView(),
        );
        if (quiz == null) {
          return Scaffold(
            appBar: AppBar(),
            drawer: drawer,
            body: Container(
              constraints: BoxConstraints.expand(),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: <Widget>[
                  CircularProgressIndicator(),
                  Text('Loading quiz...'),
                ],
              ),
            ),
          );
        }
        return WillPopScope(
          onWillPop: () async => true,
          child: Column(
            children: <Widget>[
              Expanded(
                child: PageView.builder(
                  controller: _pageController,
                  onPageChanged: (index) {
                    setState(() {
                      _currentPage = index;
                    });
                  },
                  scrollDirection: Axis.horizontal,
                  physics: PageScrollPhysics(),
                  itemCount: null,
                  itemBuilder: (BuildContext context, int index) {
                    if (index > quiz.answeredQuestionsCount) {
                      return null;
                    }
                    if (index < quiz.questionCount) {
                      if (index >= quiz.availableQuestionsCount) {
                        return Scaffold(
                          body: Container(
                            constraints: BoxConstraints.expand(),
                            child: Column(
                              mainAxisAlignment: MainAxisAlignment.center,
                              crossAxisAlignment: CrossAxisAlignment.center,
                              children: <Widget>[
                                CircularProgressIndicator(),
                                Text('Adding more questions...'),
                              ]
                            ),
                          ),
                        );
                      } else {
                        return QuestionScreen(
                          key: ObjectKey(index),
                          index: index,
                          answered: index < quiz.answeredQuestionsCount,
                          drawer: drawer,
                          onAnswer: () { _quizController.answerQuestion(index); },
                          onProceed: () { _showNextQuestion(quiz); },
                        );
                      }
                    } else {
                      return QuizResult(
                        drawer: drawer,
                      );
                    }
                  },
                ),
              ),
              if (_currentPage < quiz.questionCount) LinearProgressIndicator(
                value: quiz.answeredQuestionsCount / quiz.questionCount,
                valueColor: AlwaysStoppedAnimation(theme.primaryColorDark),
                backgroundColor: theme.backgroundColor,
              ),
            ],
          ),
        );
      },
    );
  }

  Future<void> _showNextQuestion(Quiz quiz) async {
    final targetPage = quiz.isComplete ? quiz.questionCount : quiz.answeredQuestionsCount;
    await _pageController.animateToPage(
        targetPage,
        duration: Duration(milliseconds: 400),
        curve: Curves.ease);
  }
}

class QuestionScreen extends StatelessWidget {
  final int index;
  final bool answered;
  final Drawer drawer;
  final void Function() onAnswer;
  final void Function() onProceed;

  QuestionScreen({Key key, @required this.index, @required this.answered, @required this.drawer, @required this.onAnswer, @required this.onProceed}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: !answered ? onAnswer : onProceed,
      child: Scaffold(
        extendBodyBehindAppBar: true,
        appBar: AppBar(
          backgroundColor: Colors.transparent,
          elevation: 0.0, // Prevents shadow.
          actions: <Widget>[],
        ),
        drawer: drawer,
        body: Container(
          constraints: BoxConstraints.expand(),
          color: index.isEven ? Colors.red.shade200 : Colors.yellow.shade200,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              Text('#$index', style: TextStyle(fontSize: 96.0)),
              Text(!answered ? 'Tap to answer' : 'Tap to go to next page'),
            ],
          ),
        ),
      ),
    );
  }
}

class QuizResult extends StatelessWidget {
  final Drawer drawer;

  const QuizResult({Key key, this.drawer}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('The End')),
      drawer: drawer,
      body: Container(
        constraints: BoxConstraints.expand(),
        child: Center(child: Text('Bug not reproduced :(')),
      ),
    );
  }
}
```
</details>

**Comment by darshankawar** [comment]
Thanks @ttencate for steps to replicate. I am able to consistently replicate the issue on Samsung S10+, but not on other samsung device (8.1.0).

[65747_norepro.mp4.zip](https://github.com/flutter/flutter/files/5236689/65747_norepro.mp4.zip)
[65747_s10.mp4.zip](https://github.com/flutter/flutter/files/5236692/65747_s10.mp4.zip)




<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.20.4, on Mac OS X 10.15.4 19E2269, locale en-IN)
    • Flutter version 1.20.4 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision fba99f6cf9 (2 days ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (1 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm • Android 8.1.0 (API
      27)

• No issues found!

```

</details>

**Comment by markijbema** [comment]
> Motorala Moto G6 Plus with unknown Android version: always gets stuck on the 20th bird.

Android 9, if it helps

**Comment by ttencate** [comment]
This issue is currently blocking the release of my app. I would really, really like to contribute a fix, but without being able to reproduce the issue myself, that's not really possible. If somebody is able to diagnose the issue to the point that we can repro it on all devices (maybe with a unit test?), then I would gladly take over from there!

**Comment by guillaume-fortin-keyneosoft** [comment]
@ttencate 
I reproduced your problem on a samsung galaxy A70.

the problem is line 135 inside main.dart (reproduce sample).

```
if (index > quiz.answeredQuestionsCount) {
                      return null;
 }
```

if you change code by that, your bug disappear.

```
if (index > quiz.answeredQuestionsCount) {
                   return SizedBox.shrink();
 }
```

i think that PageView.builder can't return null widget.

**Comment by ttencate** [comment]
@guillaume-fortin-keyneosoft Thank you, that's very useful!

By itself, it changes the app's behaviour in an undesirable way: you can now swipe forwards beyond the last answered question and get a blank screen, whereas previously, you'd have to answer it before you could proceed. Setting `itemCount` to a non-`null` value (i.e. `itemCount: quiz.answeredQuestionsCount + 1`) would solve that, but that caused each `QuestionScreen` to be created, destroyed and then created again for reasons I don't understand. (In my actual code, `QuestionScreen` is a `StatefulWidget` so this is problematic.) However, I can't reproduce this in the "minimal" reproduction project posted above, so that problem has its origins elsewhere in my code and can be fixed there. **In short, this gives me a viable workaround!**

However, `PageView.builder` ends up passed to `SliverChildBuilderDelegate.builder`, which [is explicitly allowed to return `null`](https://api.flutter.dev/flutter/widgets/SliverChildBuilderDelegate/builder.html). So I don't think I am doing anything wrong in my code.

Also this doesn't explain why it goes wrong on some devices but not others.

I just discovered Samsung's [Remote Test Lab](https://developer.samsung.com/remotetestlab/) and am now trying to get to the bottom of this.

**Comment by ttencate** [comment]
Here's a rather much smaller snippet that reproduces it on Samsung Galaxy S10+:

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

Future<void> main() async {
  runApp(App());
}

class App extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _AppState();
}

class _AppState extends State<App> {
  int viewedPageCount = 0;

  final _pageController = PageController(initialPage: 0);

  @override
  void dispose() {
    _pageController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: PageView.builder(
          controller: _pageController,
          itemBuilder: (BuildContext context, int index) {
            if (index > viewedPageCount) {
              return null;
            }
            return GestureDetector(
              onTap: () {
                if (index == viewedPageCount) {
                  setState(() { viewedPageCount++; });
                }
                _pageController.jumpToPage(viewedPageCount);
              },
              child: Container(
                color: index.isEven ? Colors.red.shade200 : Colors.yellow.shade200,
                child: Center(child: Text('#$index', style: TextStyle(fontSize: 96.0))),
              ),
            );
          },
        ),
      ),
    );
  }
}
```

Simply tap the screen and note how you can't get past #5 on this device. (On Moto G6, presumably you can't get past #20.)

**Comment by ttencate** [comment]
I switched to `master` and the problem disappeared!

```
Flutter 1.24.0-8.0.pre.147 • channel master • https://github.com/flutter/flutter.git
Framework • revision f04b3d917f (19 hours ago) • 2020-11-07 16:09:02 -0500
Engine • revision 0f7cdca65f
Tools • Dart 2.12.0 (build 2.12.0-27.0.dev)
```

However, it's still broken in the latest stable, 1.22.3. I'm going to bisect because I'm real curious now...

**Comment by cbracken** [comment]
Thanks for confirming @ttencate. If you narrow it down to a commit let us know!

/cc @goderbauer who would know who is the most likely person to have fixed this.

**Comment by ttencate** [comment]
And the winner is... [commit `03a3457`](https://github.com/flutter/flutter/commit/03a3457f10a18f8a024b8adc2bbd0deac0fe2443) from PR #68199 "Fix rounding issue in getMaxChildIndexForScrollOffset" which fixed #68182 "PageView adds 2 widget elements instead of 1 in the 5th slide, only in Android". So, thank you very much @daviddgl, you made my day!

**Comment by daviddgl** [comment]
> And the winner is... [commit `03a3457`](https://github.com/flutter/flutter/commit/03a3457f10a18f8a024b8adc2bbd0deac0fe2443) from PR #68199 "Fix rounding issue in getMaxChildIndexForScrollOffset" which fixed #68182 "PageView adds 2 widget elements instead of 1 in the 5th slide, only in Android". So, thank you very much @daviddgl, you made my day!

Your welcome @ttencate! 😉
